README.rst
==================================================
pyutillib - A library of general utility functions
==================================================

Python Utility Library is a small collection of functions in several categories.

Date functions
==============

In order to use a date function you need to::

    import pyutillib.date_utils

Working with date strings
-------------------------

In order to have an unambiguous relation between a date in string and datetime
format, only a limited number of formats is available. The default is often 
good enough.

Usage examples::

    >>> import pyutillib.date_utils as du
    >>> du.datestr2date('20001231')
    datetime.date(2000, 12, 31)
    >>> du.datestr2date('12/31/2000')
    datetime.date(2000, 12, 31)
    >>> du.datestr2date('31-12-2000')
    datetime.date(2000, 12, 31)

    >>> import datetime
    >>> d = datetime.date(2000, 12, 31)
    >>> du.date2datestr(d)
    '20001231'
    >>> du.date2datestr(d, 'd-m-yy')
    '31-12-00'
    >>> du.date2datestr(d, 'd-m-yyy')
    Traceback (most recent call last):
      File "<stdin>", line 1, in <module>
      File "pyutillib/date_utils.py", line 128, in date2datestr
        raise ValueError('Invalid format string, year must have 2 or 4 digits')
    ValueError: Invalid format string, year must have 2 or 4 digits
    >>> du.date2datestr(d, 'mm/dd/yyyy')
    '12/31/2000'
    >>> du.date2datestr(d, 'yymmdd')
    '001231'

Working with weekdays
---------------------

These are simple things, but hard to remember, so just a few convenience 
functions::

    >>> import pyutillib.date_utils as du
    >>> import datetime as dt
    >>> du.is_weekday(dt.date(2013,4,16))
    True
    >>> du.is_weekday(dt.date(2013,4,14))
    False
    >>> du.is_weekend(dt.date(2013,4,14))
    True
    >>> du.is_weekend(dt.date(2013,4,16))
    False
    >>> du.next_weekday(dt.date(2013,4,16))
    datetime.date(2013, 4, 17)
    >>> du.next_weekday(dt.date(2013,4,12))
    datetime.date(2013, 4, 15)
    >>> du.next_weekday(dt.date(2013,4,13))
    datetime.date(2013, 4, 15)
    >>> du.previous_weekday(dt.date(2013,4,16))
    datetime.date(2013, 4, 15)
    >>> du.previous_weekday(dt.date(2013,4,15))
    datetime.date(2013, 4, 12)
    >>> du.previous_weekday(dt.date(2013,4,14))
    datetime.date(2013, 4, 12)

Working with years
------------------

Simple, but takes leap years into account::

    >>> import pyutillib.date_utils as du
    >>> import datetime as dt
    >>> du.last_year(dt.date(2000,2,29))
    datetime.date(1999, 2, 28)
    >>> du.last_year(dt.date(2001,2,29))
    Traceback (most recent call last):
      File "<stdin>", line 1, in <module>
    ValueError: day is out of range for month
    >>> du.last_year(dt.date(2001,2,28))
    datetime.date(2000, 2, 28)
    >>> du.last_year(dt.date(2001,3,1))
    datetime.date(2000, 3, 1)

The DateList class
------------------

This class subclasses the standard Python list, so all normal list functionality
is available. The differences are that there are a few extra methods and that
the .index method works differently.

**Instantiating:** a list of dates works as expected::

    >>> import pyutillib.date_utils as du
    >>> import datetime as dt
    >>> dates = [dt.date(2012, 1, d) for d in range(1, 32)]
    >>> dl = du.DateList(dates)
    >>> dl[0]
    datetime.date(2012, 1, 1)
    >>> dl[-1]
    datetime.date(2012, 1, 31)

**index** returns the the index of a date, or (if the date is not in the list), 
the index of the most recent date before the input date::

    >>> dl.index(dt.date(2012,1,2))
    1
    >>> dl.index(dt.date(2012,1,30))
    29
    >>> dates2 = [dt.date(2012, 1, d) for d in range(1, 32, 4)]
    >>> dl2 = du.DateList(dates2)
    >>> for d in dl2: print d
    2012-01-01
    2012-01-05
    2012-01-09
    2012-01-13
    2012-01-17
    2012-01-21
    2012-01-25
    2012-01-29
    >>> dl2.index(dt.date(2012,1,8))
    1
    >>> dl2.index(dt.date(2012,1,9))
    2
    >>> dl2.index(dt.date(2012,1,10))
    2

**on_or_before** returns the input date, or (if the date is not in the list), 
the most recent date before the input date::

    >>> dl2.on_or_before(dt.date(2012,1,5))
    datetime.date(2012, 1, 5)
    >>> dl2.on_or_before(dt.date(2012,1,4))
    datetime.date(2012, 1, 1)
    >>> dl2.on_or_before(dt.date(2012,1,6))
    datetime.date(2012, 1, 5)

**delta** returns the number of days in the list between two dates::

    >>> dl.delta(dt.date(2012,1,10), dt.date(2012,1,20))
    10
    >>> dl2.delta(dt.date(2012,1,10), dt.date(2012,1,20))
    2

**offset** returns the date n_days after (or before if n_days < 0) the input
date, note that these are not calendar days, but dates in the list::

    >>> dl.offset(dt.date(2012,1,10),3)
    datetime.date(2012, 1, 13)
    >>> dl2.offset(dt.date(2012,1,10),3)
    datetime.date(2012, 1, 21)

**subset** returns a list of dates between two specified dates, only dates that
are in the original list are included::

    >>> for d in dl.subset(dt.date(2012,1,10), dt.date(2012,1,20)): print d
    2012-01-10
    2012-01-11
    2012-01-12
    2012-01-13
    2012-01-14
    2012-01-15
    2012-01-16
    2012-01-17
    2012-01-18
    2012-01-19
    2012-01-20
    >>> for d in dl2.subset(dt.date(2012,1,10), dt.date(2012,1,20)): print d
    2012-01-13
    2012-01-17

Working with time strings
-------------------------

In order to have an unambiguous relation between a time in string and 
datetime.time format, only a limited number of formats is available. The 
default is often good enough.

Usage examples::

    >>> from pyutillib import date_utils as du
    >>> du.timestr2time('123456')
    datetime.time(12, 34, 56)
    >>> du.timestr2time('12:34:56')
    datetime.time(12, 34, 56)
    >>> du.timestr2time('12:34')
    datetime.time(12, 34)
    >>> du.timestr2time('01:23')
    datetime.time(1, 23)
    >>> du.timestr2time('1:23')
    datetime.time(1, 23)
    
    >>> import datetime
    >>> t = datetime.time(23,59,59)
    >>> du.time2timestr(t)
    hhmmss 23:59:59
    '235959'
    >>> du.time2timestr(t, 'hh:mm')
    hh:mm 23:59:59
    '23:59'
    >>> du.time2timestr(t, 'hh:mm:ss')
    hh:mm:ss 23:59:59
    '23:59:59'
    >>> t = datetime.time(5,59,59)
    >>> du.time2timestr(t)
    hhmmss 05:59:59
    '055959'
    >>> du.time2timestr(t, 'hh:mm:ss')
    hh:mm:ss 05:59:59
    '05:59:59'
    >>> du.time2timestr(t, 'h:mm:ss')
    h:mm:ss 05:59:59
    '5:59:59'
    >>> du.time2timestr(t, 'h:mm')
    h:mm 05:59:59
    '5:59'
    >>> du.time2timestr(t, 'hmmss')
    hmmss 05:59:59
    Traceback (most recent call last):
      File "<stdin>", line 1, in <module>
      File "pyutillib/date_utils.py", line 383, in time2timestr
        else:
    ValueError: Invalid character in format string. The following time formats are valid:
        hhmmss
        hh:mm:ss    h:mm:ss
        hh:mm       h:mm
    Where in the latter 2 formats hh has d digits which may include a leading zero
    and h may have 1 or 2 digits and no leading zero. 
    h/hh is always in 24 hour clock.

Math functions
==============

In order to use a math function you need to::

    import pyutillib.math_utils

Safe division
-------------

Sometimes you need to divide 2 numbers without worrying about division by zero.
the ``div`` function always returns a float, even if there is an iteger result.
Also the result of 0/0 is not mathematically correct, but often practically OK.

Some examples::

    >>> from pyutillib import math_utils as mu
    >>> mu.div(1,0)
    inf
    >>> mu.div(1,1)
    1.0
    >>> mu.div(0,0)
    0.0

Evaluating conditions
---------------------

This function can be used to evaluate conditions of arbitrary complexity. The
conditions need to be in a tuple format::

    (argument1, operator, argument2)

Where either argument 1 and 2 can be booleans if operator is a logical operator
('and', 'or'), or argument 1 and 2 are python objects if operator is a comparison
operator ('lt', 'le', 'eq', 'ne', 'ge', 'gt'). In the latter case both arguments
must be comparable, i.e. be of the same type. The only exception are floats and
ints, they can be compared with each other.

The outcome of each evaluation is always a boolean and they can be nested to
any level you like, by replacing a boolean argument by another tuple, e.g.::

    >>> from pyutillib import math_utils as mu
    >>> mu.eval_conditions(((6, 'gt', 5.7), 'and', True))
    True

It is possible to provide arguments by name, if you specify their value in a
dict::

    >>> arg_dict = {'a': 11, 'b': 0.24}
    >>> condition = ('a', 'eq', 'b')
    >>> mu.eval_conditions(condition, arg_dict)
    False

String functions
================

In order to use a string function you need to::

    import pyutillib.string_utils

Generating a random string
--------------------------

This function generates a random string of specified length made up of letters
and digits. A custom character set can be specified to limit (or extend) the
collection::

    >>> from pyutillib import string_utils as su
    >>> su.random_string()
    '7xgVQZxd'
    >>> su.random_string(charset='ABC+-')
    '-+-A+CBA'
    >>> su.random_string(20, 'ABC+-')
    'A+AB--BCB++CA-A++++C'

Safely evaluating strings
-------------------------

Instead of using eval, ast provides a better (safer) alternative. This function
is just a wrapper around that function to avoid exceptions::

    >>> su.safe_eval('(2,3,4)')
    (2, 3, 4)
    >>> print su.safe_eval('import os; os.name')
    None

Working with tuples and dicts in string format
----------------------------------------------

All functions below return None if the input string does not have the required
format.

Extracting a tuple from a string::

    >>> print su.str2tuple('(1,2,3)')
    (1, 2, 3)
    >>> print su.str2tuple('[1,2,3]')
    None
    >>> print su.str2tuple('hallo')
    None

Extracting a dict from a string::

    >>> print su.str2dict('{1:2, 3:4}')
    {1: 2, 3: 4}
    >>> print su.str2dict(' {1:2, 3:4}')
    None

Getting the keys from a dict in a string. The keys will be returned in
alphabetic order::

    >>> print su.str2dict_keys('{"a":1, 2:"3", -1: 0}')
    [-1, 2, 'a']
    >>> print su.str2dict_values('{"a":1, 2:"3", -1: 0}')
    [0, '3', 1]

Translating a decimal string to an int
--------------------------------------

This function takes a string that represents a decimal number and returns an 
integer. The *decimals* argument allows you to 'shift the decimal point', e.g.::

    >>> from pyutillib import string_utils as su
    >>> su.decstr2int('123.456', 3)
    123456
    >>> su.decstr2int('123.456', 2)
    12345
    >>> su.decstr2int('123.456', 4)
    1234560
    >>> su.decstr2int('123', 4)
    1230000
    >>> su.decstr2int('123.456', -1)
    12

Note from the above examples that the input string does not need to contain a 
decimal point and also the decimals argument may be negative (or 0).


CHANGES.txt
v0.1.0, 20130416 -- Initial release.
v0.1.1, 20130416 -- updated README
v0.2.0, 20130424 -- added DateList class
v0.3.0, 20130501 -- added time2timestr, timestr2time and decstr2int functions


pyutillib/math_utils.py
'''
pyutillib/math_utils.py

Copyright (C) 2013 Edwin van Opstal

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU Affero General Public License as
published by the Free Software Foundation, either version 3 of the
License, or (at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU Affero General Public License for more details.

You should have received a copy of the GNU Affero General Public License
along with this program.  If not, see `<http://www.gnu.org/licenses/>`.
'''

from __future__ import division
from __future__ import absolute_import

import operator
from pyutillib.string_utils import str2tuple


def div(numerator, denominator):
    '''
    Returns numerator / denominator, but instead of a ZeroDivisionError:
        0 / 0 = 0.
        x / 0 = float('inf')
    This is not mathematically correct, but often practically OK.

    Args:
        numerator (float or int)
        denominator (float or int)
    Returns:
        (float)
    Raises:
        -
    '''
    try:
        return numerator/denominator
    except ZeroDivisionError:
        if numerator == 0:
            return 0.
        elif denominator == 0:
            return float('inf')
#            return None
        else:
            return numerator/denominator


def eval_conditions(conditions=None, data={}):
    '''
    Evaluates conditions and returns Boolean value.

    Args:
        conditions (tuple) for the format of the tuple, see below
        data (dict) the keys of which can be used in conditions
    Returns:
        (boolea)
    Raises:
        ValueError if an invalid operator value is specified
        TypeError if: 
                conditions are not a 3-item tuple
                the arguments of the condition don't have the same type
                    e.g. ('abc', 'eq', 3)
                if a boolean operator does not get boolean arguments
                    e.g. (True, 'and', 15)

    The format of the condition tuple is:
        (arg1, op, arg2)
    where:
        arg1, arg2 can be numerical values, strings or condition tuples
        op is a valid operator from the operator module
    If arg is a string, and the string is a key in <data> it is treated as
    a variable with value data[arg].

    Notes:
        * If no conditions are specified True is returned.
        * empty or 0 values do *not* evaluate to booleans
    '''
#CONSIDER: implementing addition/subtraction/multiplication/division
    if not conditions:
        return True
    if isinstance(conditions, str) or isinstance(conditions, unicode):
        conditions = str2tuple(conditions)
    if not isinstance(conditions, tuple) or not len(conditions) == 3:
        raise TypeError('conditions must be a tuple with 3 items.')
    arg1 = conditions[0]
    op = conditions[1]
    arg2 = conditions[2]
    if arg1 in data:
        arg1 = data[arg1]
    elif isinstance(arg1, tuple):
        arg1 = eval_conditions(arg1, data)
    if arg2 in data:
        arg2 = data[arg2]
    elif isinstance(arg2, tuple):
        arg2 = eval_conditions(arg2, data)
    if op in ('lt', 'le', 'eq', 'ne', 'ge', 'gt'):
        if not (type(arg1) in (float, int) and type(arg2) in (float,int)) and \
                type(arg1) != type(arg2):
            raise TypeError('both arguments must have the same type {}, {}'.\
                    format(arg1, arg2))
    elif op in ('and', 'or'):
        if not isinstance(arg1, bool) or not isinstance(arg2, bool):
            raise TypeError('boolean operator {} needs boolean arguments {},'\
                    ' {}'.format(op, arg1, arg2))
        op += '_'
    else:
        raise ValueError('operator {} not supported', op)
    return getattr(operator, op)(arg1, arg2)


pyutillib/string_utils.py
'''
pyutillib/string_utils.py

Copyright (C) 2013 Edwin van Opstal

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU Affero General Public License as
published by the Free Software Foundation, either version 3 of the
License, or (at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU Affero General Public License for more details.

You should have received a copy of the GNU Affero General Public License
along with this program.  If not, see `<http://www.gnu.org/licenses/>`.
'''

from __future__ import division
from __future__ import absolute_import

import ast
import random
import string


def random_string(length=8, charset=None):
    '''
    Generates a string with random characters. If no charset is specified, only
    letters and digits are used.

    Args:
        length (int) length of the returned string
        charset (string) list of characters to choose from
    Returns:
        (str) with random characters from charset
    Raises:
        -
    '''
    if length < 1:
        raise ValueError('Length must be > 0')
    if not charset:
        charset = string.letters + string.digits
    return ''.join(random.choice(charset) for unused in xrange(length))


def safe_eval(str_in):
    '''
    Extracts a python object from a string.

    Args:
        str_in (string) that contains python variable
    Returns:
        (object) of standard python type or None of no valid object was found.
    Raises:
        -
    '''
    try:
        return ast.literal_eval(str_in)
    except:
        return None


def str2dict(str_in):
    '''
    Extracts a dict from a string.

    Args:
        str_in (string) that contains python dict
    Returns:
        (dict) or None if no valid dict was found
    Raises:
        -
    '''
    dict_out = safe_eval(str_in)
    if not isinstance(dict_out, dict):
        dict_out = None
    return dict_out


def str2tuple(str_in):
    '''
    Extracts a tuple from a string.

    Args:
        str_in (string) that contains python tuple
    Returns:
        (dict) or None if no valid tuple was found
    Raises:
        -
    '''
    tuple_out = safe_eval(str_in)
    if not isinstance(tuple_out, tuple):
        tuple_out = None
    return tuple_out


#used to be get_dict_keys
def str2dict_keys(str_in):
    '''
    Extracts the keys from a string that represents a dict and returns them
    sorted by key.

    Args:
        str_in (string) that contains python dict
    Returns:
        (list) with keys or None if no valid dict was found
    Raises:
        -
    '''
    tmp_dict = str2dict(str_in)
    if tmp_dict is None:
        return None
    return sorted([k for k in tmp_dict])


#used to be get_dict_values
def str2dict_values(str_in):
    '''
    Extracts the values from a string that represents a dict and returns them
    sorted by key.

    Args:
        str_in (string) that contains python dict
    Returns:
        (list) with values or None if no valid dict was found
    Raises:
        -
    '''
    tmp_dict = str2dict(str_in)
    if tmp_dict is None:
        return None
    return [tmp_dict[key] for key in sorted(k for k in tmp_dict)]


def decstr2int(dec_str, decimals):
    '''
    Returns an integer that has the value of the decimal string:
        dec_str*10^decimals

    Arguments:
        dec_str (string) that represents a decimal number
        decimals (int): number of decimals for creating the integer output
    Returns:
        (int)
    Raises:
        ValueError if dec_string is not a valid decimal string
        TypeError if decimals is not an integer
    Note: values may be truncated (not rounded).
    '''
    if not isinstance(decimals, int):
        raise TypeError('decimals must be an integer')
    try:
        dollars, cents = dec_str.split('.')
    except ValueError:
        if '.' not in dec_str:
            dollars = dec_str
            cents = '0'
        else:
            raise ValueError('Invalid decimal string')
    else:
        if len(cents) < decimals:
            cents = cents.ljust(decimals, '0')
        elif decimals < 1:
            cents = '0'
        elif len(cents) > decimals:
            cents = cents[:decimals]
    try:
        cents = int(cents)
    except:
        cents = 0
    try:
        return int(int(dollars) * (10 ** decimals)) + cents
    except:
        raise ValueError('Invalid decimal string')


pyutillib/date_utils.py
'''
pyutillib/date_utils.py

Copyright (C) 2013 Edwin van Opstal

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU Affero General Public License as
published by the Free Software Foundation, either version 3 of the
License, or (at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU Affero General Public License for more details.

You should have received a copy of the GNU Affero General Public License
along with this program.  If not, see `<http://www.gnu.org/licenses/>`.
'''

from __future__ import division
from __future__ import absolute_import

import datetime


VALID_DATE_FORMATS_TEXT = '''The following date formats are valid:
    yymmdd    yyyymmdd
    d-m-yy    d-m-yyyy
    m/d/yy    m/d/yyyy
Where in the latter 4 formats m and d may be 1 or 2 digits which may include
a leading zero.
In case of a 2-digit year, it is assumed to be after 2000'''


def datestr2date(date_str):
    '''
    Turns a string into a datetime.date object. This will only work if the 
    format can be "guessed", so the string must have one of the formats from
    VALID_DATE_FORMATS_TEXT.

    Args:
        date_str (str) a string that represents a date
    Returns:
        datetime.date object
    Raises:
        ValueError if the input string does not have a valid format.
    '''
    if any(c not in '0123456789-/' for c in date_str):
        raise ValueError('Illegal character in date string')
    if '/' in date_str:
        try:
            m, d, y = date_str.split('/')
        except:
            raise ValueError('Date {} must have no or exactly 2 slashes. {}'.
                    format(date_str, VALID_DATE_FORMATS_TEXT))
    elif '-' in date_str:
        try:
            d, m, y = date_str.split('-')
        except:
            raise ValueError('Date {} must have no or exactly 2 dashes. {}'.
                    format(date_str, VALID_DATE_FORMATS_TEXT))
    elif len(date_str) == 8 or len(date_str) == 6:
        d = date_str[-2:]
        m = date_str[-4:-2]
        y = date_str[:-4]
    else:
        raise ValueError('Date format not recognised. {}'.format(
                VALID_DATE_FORMATS_TEXT))
    if len(y) == 2:
        year = 2000 + int(y)
    elif len(y) == 4:
        year = int(y)
    else:
        raise ValueError('year must be 2 or 4 digits')
    for s in (m, d):
        if 1 <= len(s) <= 2:
            month, day = int(m), int(d)
        else:
            raise ValueError('m and d must be 1 or 2 digits')
    try:
        return datetime.date(year, month, day)
    except ValueError:
        raise ValueError('Invalid date {}. {}'.format(date_str, 
                VALID_DATE_FORMATS_TEXT))


def date2datestr(date, fmt='yyyymmdd'):
    '''
    Turns a datetime.date object into a string. The string must have one of the
    formats from VALID_DATE_FORMATS_TEXT to make it compatible with 
    datestr2date.

    Args:
        date (datetime.date) the date to be translated
        fmt (str) a format string.
    Returns:
        (str) that represents a date.
    Raises:
        ValueError if the format is not valid.
    '''
    if '-' in fmt:
        if not fmt.index('d') < fmt.index('m') < fmt.index('y'):
            raise ValueError('Invalid format string. {}'.format(
                    VALID_DATE_FORMATS_TEXT))
        d, m, y = fmt.split('-')
    elif '/' in fmt:
        if not fmt.index('m') < fmt.index('d') < fmt.index('y'):
            raise ValueError('Invalid format string. {}'.format(
                    VALID_DATE_FORMATS_TEXT))
        m, d, y = fmt.split('/')
    elif any(c not in 'dmy' for c in fmt):
        raise ValueError('Invalid character in format string. {}'.format(
                VALID_DATE_FORMATS_TEXT))
    else:
        if not fmt.index('y') < fmt.index('m') < fmt.index('d'):
            raise ValueError('Invalid format string. {}'.format(
                    VALID_DATE_FORMATS_TEXT))
        y, m, d = fmt[:-4], fmt[-4:-2], fmt[-2:]
    for string, char in ((d, 'd'), (m, 'm'), (y, 'y')):
        if any(c != char for c in string):
            raise ValueError('Invalid date format: {} is not {}'.\
                    format(char, string))
    if len(y) == 4:
        fmt = fmt.replace('yyyy', '%Y', 1)
    elif len(y) == 2:
        fmt = fmt.replace('yy', '%y', 1)
    else:
        raise ValueError('Invalid format string, year must have 2 or 4 digits')
    if len(m) == 2:
        fmt = fmt.replace('mm', '%m', 1)
    elif len(m) == 1:
        fmt = fmt.replace('m', 'X%m', 1)
    else:
        raise ValueError('Invalid format string, month must have 1 or 2 digits')
    if len(d) == 2:
        fmt = fmt.replace('dd', '%d', 1)
    elif len(d) == 1:
        fmt = fmt.replace('d', 'X%d', 1)
    else:
        raise ValueError('Invalid format string, day must have 1 or 2 digits')
    return date.strftime(fmt).replace('X0','X').replace('X','')


def is_weekday(date):
    '''
    Returns a boolean that indicates if date is a weekday.

    Args:
        date (datetime or datetime.date)
    Returns:
        (boolean)
    Raises:
        -
    '''
    return True if date.weekday() < 5 else False


def is_weekend(date):
    '''
    Returns a boolean that indicates if date is in a weekend.

    Args:
        date (datetime or datetime.date)
    Returns:
        (boolean)
    Raises:
        -
    '''
    return not is_weekday(date)


def previous_weekday(date):
    '''
    Returns the last weekday before date

    Args:
        date (datetime or datetime.date)
    Returns:
        (datetime or datetime.date)
    Raises:
        -
    '''
    weekday = date.weekday()
    if weekday == 0:
        n_days = 3
    elif weekday == 6:
        n_days = 2
    else:
        n_days = 1
    return date - datetime.timedelta(days=n_days)


def next_weekday(date):
    '''
    Return the first weekday after date

    Args:
        date (datetime or datetime.date)
    Returns:
        (datetime or datetime.date)
    Raises:
        -
    '''
    n_days = 7 - date.weekday()
    if n_days > 3:
        n_days = 1
    return date + datetime.timedelta(days=n_days)


def last_year(date_):
    '''
    Returns the same date 1 year ago.

    Args:
        date (datetime or datetime.date)
    Returns:
        (datetime or datetime.date)
    Raises:
        -
    '''
    day = 28 if date_.day == 29 and date_.month == 2 else date_.day
    return datetime.date(date_.year-1, date_.month, day)


class DateList(list):
    '''
    Provides a list of dates with methods to extract information.
    The dates list MUST be sorted for the methods of this class to work.
    '''

    ONE_DAY = datetime.timedelta(days=1)

    def __init__(self, dates, sort=True):
        '''
        Constructor stores the list of dates. The list will be sorted by
        default, you can avoid sorting overhead if you are 100% sure the dates
        list is sorted by specifying sort=False.
        '''
        if sort:
            dates.sort()
        list.__init__(self, dates)

    def index(self, date):
        '''
        Overloads the default list.index, because of special behaviour if the
        date is not in the list:
            - If <date> is not in the list, the index of the latest date before
                <date> is returned.
            - If <date> is earlier than the earliest date in the list, the 
                return value is 0
            - If <date> is later than the latest date in self.dates the return
                value is the index of the most recent date.
        '''
        if date in self:
            index = super(DateList, self).index(date)
        elif date < self[0]:
            index = 0
        elif date > self[-1]:
            index = len(self) - 1
        else:
            while date not in self:
                date -= self.ONE_DAY
            index = super(DateList, self).index(date)
        return index

#    def latest_date_before(self, date):
    def on_or_before(self, date):
        '''
        returns the latest date from self.dates that is <= <date>
        '''
        return self[self.index(date)]

    def delta(self, fromdate, todate):
        '''
        Return the number of dates in the list between <fromdate> and
        <todate>.
        '''
#CONSIDER: raise an exception if a date is not in self
        return self.index(todate) - self.index(fromdate)

    def offset(self, date, n_days):
        '''
        Return the date n_days after (or before if n_days < 0) <date>
        note: not calender days, but self days!
        '''
#CONSIDER: raise an exception if a date is not in self
        index = self.index(date) + n_days
        if index < 0:
            index = 0
        if index > len(self) - 1:
            index = len(self) - 1
        return self[index]

    def subset(self, fromdate, todate):
        '''
        Return a list of dates from the list between <fromdate> and <todate>
        (inclusive)
        '''
#CONSIDER: raise an exception if a date is not in self
        i_from = self.index(fromdate)
        if fromdate != self[i_from]:
            # don't go back before fromdate
            i_from += 1
        i_to = self.index(todate)
        return self[i_from:i_to + 1]


VALID_TIME_FORMATS_TEXT = '''The following time formats are valid:
    hhmmss
    hh:mm:ss    h:mm:ss
    hh:mm       h:mm
Where in the latter 2 formats hh has d digits which may include a leading zero
and h may have 1 or 2 digits and no leading zero. 
h/hh is always in 24 hour clock.
'''

def timestr2time(time_str):
    '''
    Turns a string into a datetime.time object. This will only work if the 
    format can be "guessed", so the string must have one of the formats from
    VALID_TIME_FORMATS_TEXT.

    Args:
        time_str (str) a string that represents a date
    Returns:
        datetime.time object
    Raises:
        ValueError if the input string does not have a valid format.
    '''
    if any(c not in '0123456789:' for c in time_str):
        raise ValueError('Illegal character in time string')
    if time_str.count(':') == 2:
        h, m, s = time_str.split(':')
    elif time_str.count(':') == 1:
        h, m = time_str.split(':')
        s = '00'
    elif len(time_str) == 6:
        h = time_str[:2]
        m = time_str[2:4]
        s = time_str[4:]
    else:
        raise ValueError('Time format not recognised. {}'.format(
                VALID_TIME_FORMATS_TEXT))
    if len(m) == 2 and len(s) == 2:
        mins = int(m)
        sec = int(s)
    else:
        raise ValueError('m and s must be 2 digits')
    try:
        return datetime.time(int(h), mins, sec)
    except ValueError:
        raise ValueError('Invalid time {}. {}'.format(time_str, 
                VALID_TIME_FORMATS_TEXT))

def time2timestr(time, fmt='hhmmss'):
    '''
    Turns a datetime.time object into a string. The string must have one of the
    formats from VALID_TIME_FORMATS_TEXT to make it compatible with 
    timestr2time.

    Args:
        time (datetime.time) the time to be translated
        fmt (str) a format string.
    Returns:
        (str) that represents a time.
    Raises:
        ValueError if the format is not valid.
    '''
    if fmt.count(':') == 2:
        if not fmt.index('h') < fmt.index('m') < fmt.index('s'):
            raise ValueError('Invalid format string. {}'.format(
                    VALID_TIME_FORMATS_TEXT))
        h, m, s = fmt.split(':')
    elif fmt.count(':') == 1:
        if not fmt.index('h') < fmt.index('m'):
            raise ValueError('Invalid format string. {}'.format(
                    VALID_TIME_FORMATS_TEXT))
        h, m = fmt.split(':')
        s = None
    elif any(c not in 'hms' for c in fmt) or len(fmt) != 6:
        raise ValueError('Invalid character in format string. {}'.format(
                VALID_TIME_FORMATS_TEXT))
    else:
        if not fmt.index('h') < fmt.index('m') < fmt.index('s'):
            raise ValueError('Invalid format string. {}'.format(
                    VALID_TIME_FORMATS_TEXT))
        h, m, s = fmt[:-4], fmt[-4:-2], fmt[-2:]
    for string, char in ((h, 'h'), (m, 'm'), (s, 's')):
        if string is not None and any(c != char for c in string):
            raise ValueError('Invalid date format: {} is not {}'.\
                    format(char, string))
    if len(h) == 2:
        fmt = fmt.replace('hh', '%H', 1)
    elif len(h) == 1:
        fmt = fmt.replace('h', 'X%H', 1)
    else:
        raise ValueError('Invalid format string, hour must have 1 or 2 digits')
    if len(m) == 2:
        fmt = fmt.replace('mm', '%M', 1)
    else:
        raise ValueError('Invalid format string, minutes must have 2 digits')
    if s is not None and len(s) == 2:
        fmt = fmt. replace('ss', '%S', 1)
    elif s is not None:
        raise ValueError('Invalid format string, seconds must have 2 digits')
    return time.strftime(fmt).replace('X0','X').replace('X','')

